smart-contracts/contracts/cl-vault/src/vault/swap.rs

```diff
@@ -84,20 +99,129 @@ pub fn swap(
     // we will only ever have a route length of one, this will likely change once we start selecting different routes
     let pool_route = SwapAmountInRoute {
         pool_id: pool_config.pool_id,
-        token_out_denom,
+        token_out_denom: token_out_denom.to_string(),
     };
 
-    let swap_msg: CosmosMsg =
-        osmosis_std::types::osmosis::poolmanager::v1beta1::MsgSwapExactAmountIn {
-            sender: env.contract.address.to_string(),
-            routes: vec![pool_route],
-            token_in: Some(OsmoCoin {
-                denom: token_in_denom.to_string(),
-                amount: token_in_amount.to_string(),
-            }),
-            token_out_min_amount: token_out_min_amount.to_string(),
+    let swap_msg: Result<CosmosMsg, _> = match dex_router {
+        Some(dex_router_address) => {
+            let offer_asset = AssetInfo::Native(token_in_denom.to_string());
+            let ask_asset = AssetInfo::Native(token_out_denom.to_string());
+
+            let recommended_out: Uint128 = match recommended_swap_route.clone() {
+                Some(operations) => deps.querier.query_wasm_smart(
+                    dex_router_address.to_string(),
+                    &QueryMsg::SimulateSwapOperations {
+                        offer_amount: token_in_amount,
+                        operations,
+                    },
+                )?,
+                None => 0u128.into(),
+            };
+            let best_path: Option<BestPathForPairResponse> = deps.querier.query_wasm_smart(
+                dex_router_address.to_string(),
+                &QueryMsg::BestPathForPair {
+                    offer_asset: offer_asset.into(),
+                    ask_asset: ask_asset.into(),
+                    exclude_paths: None,
+                    offer_amount: token_in_amount,
+                },
+            )?;
+            let best_out = match best_path.clone() {
+                Some(best_path) => best_path.return_amount,
+                None => 0u128.into(),
+            };
+
+            // if we need to force the route
+            if force_swap_route {
```

Comment:
this if else chain is relatively hard to read in it's current form

Reply:
Perhaps a better way to do this function would be to nest some if/else statements for eveluating different paths. If I understand this function correctly, it could be (in pseudocode)
```
let swap_route = if force_swap_route {
    forced_swap_route
} else {
    deps.querier.query_wasm_smart(
                dex_router_address.to_string(),
                &QueryMsg::BestPathForPair {
                    offer_asset: offer_asset.into(),
                    ask_asset: ask_asset.into(),
                    exclude_paths: None,
                    offer_amount: token_in_amount,
                },
            )?;
}

Ok( execute_swap_operations(
                    dex_router_address,
                    recommended_swap_route.unwrap(), // will be some here
                    token_out_min_amount,
                    token_in_denom,
                    token_in_amount,
                ))
```

instead of the way it's currently written

Reply:
Noted.